### PR TITLE
mon: mark_down_pgs in lockstep with pg_map's osdmap epoch

### PR DESCRIPTION
**Pull request by liewegas**
This fixes the stale PGs we've been seeing.  And probably
several corner cases that are hard to see.

### CODE REVIEW

**Comment by liewegas** [comment]
@tchaikov @jecluis review please?

**Comment by liewegas** [comment]
passed rados suite

**Comment by liewegas** [comment]
added a fix for a race between map_pg_create and a pg_stat update, observed at http://pulpito.ceph.com/sage-2016-03-20_11:34:21-rados-wip-sage-testing-distro-basic-smithi/76029

File: src/mon/PGMonitor.cc [hunk 1]
```diff
@@ -1310,9 +1310,19 @@ void PGMonitor::_try_mark_pg_stale(
...
+	   << pg_map.last_osdmap_epoch << dendl;
+  if (pg_map.last_osdmap_epoch == 0)
+    return false;
+
+  // use the OSDMap that matches the one pg_map has consumed.
+  std::unique_ptr<OSDMap> osdmap;
+  bufferlist bl;
+  int err = mon->osdmon()->get_version_full(pg_map.last_osdmap_epoch, bl);
+  assert(err == 0);
+  osdmap.reset(new OSDMap);
```
Comment (tchaikov) [inline src/mon/PGMonitor.cc]: could use `std::make_unique()`.

File: src/mon/PGMonitor.cc [hunk 2]
```diff
@@ -1310,9 +1310,19 @@ void PGMonitor::_try_mark_pg_stale(
...
-  dout(10) << "check_down_pgs" << dendl;
+  dout(10) << "check_down_pgs last_osdmap_epoch "
+	   << pg_map.last_osdmap_epoch << dendl;
+  if (pg_map.last_osdmap_epoch == 0)
+    return false;
+
+  // use the OSDMap that matches the one pg_map has consumed.
+  std::unique_ptr<OSDMap> osdmap;
+  bufferlist bl;
+  int err = mon->osdmon()->get_version_full(pg_map.last_osdmap_epoch, bl);
```
Comment (tchaikov) [inline src/mon/PGMonitor.cc]: the OSDMonitor kicks pgmon in `OSDMonitor::update_from_paxos()`, so the pgmon can consume (`check_osd_map()`) the latest osdmaps once the monitor is **active**[1]. and pg_map.last_osdmap_epoch is updated in `PGMonitor::update_from_paxo()` when the monitor is still **updating**.

so, is there a chance that `check_down_pgs()` gets called **after** `pg_map.last_osdmap_epoch` is updated, and **before** `check_osd_map()` is called? in other words:

1. pg_map.set_last_osdmap_epoch(last_osdmap_epoch);
1. PGMonitor::check_down_pgs()
1. check_osd_map()

---
[1] see `PGMonitor::check_osd_map()`
```c++
  if (!is_writeable()) {
    dout(10) << __func__ << " -- pgmap not writeable, waiting" << dendl;
    wait_for_writeable_ctx(new RetryCheckOSDMap(this, epoch));
    return;
  }
```

the monitor is not writable until it is done with refresh (the `update_from_paxos()` calls).

Comment (tchaikov) [inline src/mon/PGMonitor.cc]: please ignore the comment above, just read b67ac8c318f8bf635b221b626040b802d2f119c2 , which renders the case above possible

**Comment by tchaikov** [comment]
lgtm modulo the suggestion to use `std::make_unique()`

**Comment by liewegas** [comment]
will do one more rados run with the new commit

**Pull request merged**